### ISSUE TITLE
Wins to PASS after match leaves queue don't promote?

### ISSUE THREAD

**Issue by roy7** [gcp/leela-zero-server#63]
See discussion at https://github.com/gcp/leela-zero/issues/1229#issuecomment-383298609 and https://github.com/gcp/leela-zero/issues/1228#issue-316454330.

It appears a network that had sent out all pages and left pending matches queue had enough wins arrive to show PASS on the web site, but it didn't actually promote. The same SPRT()===true test is used to promote a late arrival match win and to decide to show PASS on web site. So I wonder if this is a bug that matches after the match has ended can't promote a network for some reason?

Just looking at https://github.com/gcp/leela-zero-server/blob/next/server.js#L883 I don't see why it would fail however.

**Comment by rchs819** [comment]
Here is the replay of the [match](http://zero.sjeng.org/match-games/5ada73bae288f3147755b60c) for `59bb7337` vs `e8601c38`

Let's start look from Game 395 to 400 onwards

Game #| Win / Loss| SPRT|Note
-- | -- | --| --
395 | 216:179 (54.68%) | null
396 | 217:179 (54.80%) | null
397 | 217:180 (54.66%) | null
398 | 218:180 (54.77%) | null
399 | 218:181 (54.64%) | null
400 | 219:181 (54.75%) | FALSE| Failed at 400th game, so network `59bb7337` is out
401 | 220:181 (54.86%) | FALSE
402 | 220:182 (54.73%) | FALSE
403 | 221:182 (54.84%) | FALSE
404 | 222:182 (54.95%) | FALSE
405 | 223:182 (55.06%) | TRUE | made a come back from later matches, but too late
406 | 223:183 (54.93%) | FALSE
407 | 223:184 (54.79%) | FALSE
408 | 223:185 (54.66%) | FALSE
409 | 223:186 (54.52%) | FALSE
410 | 223:187 (54.39%) | FALSE
411 | 223:188 (54.26%) | FALSE
412 | 224:188 (54.37%) | FALSE
413 | 224:189 (54.24%) | FALSE
414 | 224:190 (54.11%) | FALSE
415 | 225:190 (54.22%) | FALSE
416 | 225:191 (54.09%) | FALSE
417 | 226:191 (54.20%) | FALSE
418 | 227:191 (54.31%) | FALSE
419 | 227:192 (54.18%) | FALSE
420 | 228:192 (54.29%) | FALSE
421 | 229:192 (54.39%) | FALSE
422 | 230:192 (54.50%) | FALSE
423 | 231:192 (54.61%) | FALSE
424 | 232:192 (54.72%) | FALSE
425 | 233:192 (54.82%) | FALSE
426 | 234:192 (54.93%) | FALSE
427 | 235:192 (55.04%) | TRUE
428 | 235:193 (54.91%) | FALSE
429 | 235:194 (54.78%) | FALSE
430 | 236:194 (54.88%) | FALSE
431 | 237:194 (54.99%) | FALSE
432 | 237:195 (54.86%) | FALSE
433 | 238:195 (54.97%) | FALSE
434 | 239:195 (55.07%) | TRUE
435 | 239:196 (54.94%) | FALSE
436 | 240:196 (55.05%) | TRUE
437 | 240:197 (54.92%) | FALSE
438 | 240:198 (54.79%) | FALSE
439 | 240:199 (54.67%) | FALSE

**Comment by roy7** [comment]
So this confirms the bug. 4 times we saw a PASS and https://github.com/gcp/leela-zero-server/blob/next/server.js#L883 should have promoted the network.

In the mean time, @gcp should we manually promote this?

**Comment by roy7** [comment]
I gotta run for the day but someone look at https://github.com/gcp/leela-zero-server/blob/bbab4f7fe280b3ae8b1388d2afc8af8372d9a953/server.js#L301. Does being over max cause problems in here? @Dorus? Might the "can reach" behave poorly if we're already over the max and perhaps is saying FAIL even though it's actually a PASS?

But the SPRT() is called to print "PASS" on the web site and that call isn't failing...

**Comment by rchs819** [comment]
@roy7 
For submission Game 405

Game #| Win / Loss| SPRT
-- | -- | --
404 | 222:182 (54.95%) | FALSE
405 | 223:182 (55.06%) | TRUE 

We expect the network to be promoted from the line below

https://github.com/gcp/leela-zero-server/blob/bbab4f7fe280b3ae8b1388d2afc8af8372d9a953/server.js#L883-L886

However, according to my debugging result, I `console.log(match)` just before the `if` statement.
The log shows that the `match` has the value of
```
{
  ...
  network1_losses: 182,
  network1_wins: 222,
  game_count: 404,
  ...
}
```
which means the submission result hasn't been written to database yet, hence the promotion check didn't go through.

**Comment by Dorus** [comment]
@roy7 This line should kick in: https://github.com/gcp/leela-zero-server/blob/next/server.js#L304

```js
if(w+l>=max&&w/(w+l)>=(aim/max)) return true;
```
So let
```text
w=223
l=182
max=400
aim = max / 2 + 2 * stDev(max)=220
```
then
```text
w+l>=max&&w/(w+l)>=(aim/max)
405>=400 && 0.5506 >=0.55
true && true
true
```
QED

I tried this in console and JS agrees with me, so this line isn't the problem. Must be elsewhere. I'll keep looking.

**Comment by roy7** [comment]
@rchs819 Ahhhh!

> which means the Game 405 submission result hasn't been written to database yet, hence the promotion check didn't go through.

So maybe it's a race condition and we aren't calling this in the callback properly?

In this case, the next match that arrives would see right info in the database, but the check isn't done on a loss. So we'd need 2 match wins in a row to see the result of the prior match win being saved for this to work correctly.

If it isn't a race condition then we need to add 1 to the wins count and games count before doing the tests, but I'm guessing it's a race with the recording of the result and we need to do this test after the database update completes?

**Comment by roy7** [comment]
I vaguely have a memory of this happening before but I thought we fixed it already previously.

**Comment by Dorus** [comment]
If we only need one additional win, I'm not too worried. Still something worth fixing but nothing major. I was worried any promotion above 400 was impossible and that would have been bad.

Possibly we could delay sprt fail above 400 if it's really close to 55%? That way the next match would also be against current best. If we base it on still ongoing matches it should eventually go to pass / fail

**Comment by roy7** [comment]
Man I'm sure we discussed this situation a long time ago. I guess it didn't get fixed or I did something else that I thought fixed it and was wrong. Maybe it was more of a TODO than I never went back to.

But final check for promotion for a network no longer in the pending match queue needs to be in the callback for https://github.com/gcp/leela-zero-server/blob/c6dcf1ba8310512f72654e00327ae4cd3ed33374/server.js#L769 someplace so the database is always updated before we check for this situation.

**Comment by LRGB** [comment]
Just to note that some late matches have come in and now 59bb7337 is registering as a PASS with 55.18%

**Comment by ryouiki** [comment]
@LRGB It was not late matches. Kind of fabricated attack to promote certain network. See #60

**Comment by roy7** [comment]
Correct, that client was never sent match requests for those games per the logs. The only time their IP address appears is uploading those 5 match results.

**Comment by Mardak** [comment]
This should be fixed with #64 so two wins in a row are no longer required.
